django/http/response.py

```diff
@@ -397,14 +400,39 @@ def __init__(self, *args, as_attachment=False, filename='', **kwargs):
         self.filename = filename
         super().__init__(*args, **kwargs)
 
+    def _wrap_file_to_stream_close(self, filelike):
+        """
+        Wrap the file-like close() with a version that calls
+        FileResponse.close().
+        """
+        closing = False
+        filelike_close = getattr(filelike, 'close', lambda: None)
+
+        def file_wrapper_close():
+            nonlocal closing
+            # Prevent an infinite loop since FileResponse.close() tries to
+            # close the objects in self._closable_objects.
+            if closing:
+                return
+            closing = True
+            try:
+                filelike_close()
+            finally:
+                self.close()
```

Comment:
The filelike object will be tried closed twice. since it's part of `_closable_objects`

Reply:
Good question. That's okay because that's what the first three lines of the wrapped `file_wrapper_close()` are for:

```python
if closing:
    return
closing = True
```

If `filelike.close()` is called a second time, it will immediately return the second time.

The reason we need to add `filelike` to `_closable_objects` at all is we don't know in advance if `wsgi.file_wrapper` will be used or not. In the case that `wsgi.file_wrapper` isn't used, being in `_closable_objects` ensures the `filelike` will be closed. This is because the WSGI server will call close on the `Response` object rather than on the `filelike` object, and `Response.close()` is implemented in Django as iterating through the objects in `_closable_objects` and calling `close()` on them.

Reply:
What's the case? (Alternatively, does the `closing` guard not cover this?)

Reply:
@cjerdonek good point, I see the issue with it being either called by the `file_wrapper` or by the Django response. I guess that is the best solution then.

I would suggest going nuts on comments here. I will have certainly forgotten all about this code in a month or two ;)

Reply:
@codingjoe Thanks! I'm happy to add more, though I'll defer to the Django committers here for now as e.g. @felixxm has already done a pass-through of his own clean-ups, etc.

Also, note there is already a 2-line comment at the beginning of `file_wrapper_close()`, which was meant to cover the issue you mentioned:

```python
# Prevent an infinite loop since FileResponse.close() tries to
# close the objects in self._closable_objects.
```

Reply:
OK , thanks @codingjoe, @cjerdonek both. On reflection, I think it's OK as-is.